specs/ChainSpec/Consensus.md

```diff
@@ -0,0 +1,147 @@
+# Consensus
+
+The fields in the Block header relevant to the consensus process are:
+
+```rust
+struct BlockHeader {
+    ...
+    prev_hash: BlockHash,
+    height: BlockHeight,
+    epoch_id: EpochId,
+    last_final_block_hash: BlockHash,
+    approvals: Vec<Approval>
+    ...
+}
+```
+
+Each block belongs to a particular epoch, and has a particular height. The highest height block known to a particular node is its head.
+
+Each epoch has a set of block producers who are assigned to produce blocks in the epoch, and has an assigned block producer for each height in the epoch. The block producers and the assignments in a particular epoch are known from the first block of the *preceding* epoch.
+
+Consecutive blocks do not necessarily have sequential heights. For example, if a block producer responsible for producing a block at height `h` is offline, the block at height `h+1` can have as its previous block a block with height `h-1`.
+
+If a block `B` at some height `h` builds on top of a block `B_prev` that has height `h-1`, and `B_prev` in turn builds on top of a block `B_prev_prev` that has height `h-2` (i.e. if the last three blocks in the chain ending at `B` have sequential heights), and all three blocks are in the same epoch, the block `B_prev_prev` is final, and cannot be reverted unless block producers with more than `2/3` cumulative stake deviate from the protocol. 
+
+TODO: talk about epoch boundaries here
+
+Block producers in the particular epoch exchange many kinds of messages. The two kinds that are relevant to the consensus are **Blocks** and **Approvals**. The approval contains five fields:
+
+```rust
+struct Approval {
+    prev_hash: BlockHash,
+    prev_height: BlockIndex,
```

Comment:
Why do we need `prev_height`? It can be derived.

Reply:
Even though not part of the spec, each implementation will need to have some sort of cache for the approvals on blocks that we haven't seen yet. Since the signature on the skip is on (prev_height, target_height), you won't be able to verify the signature for an approval that arrived before the block. I don't think a non-obusable approval cache is possible in such a case.